compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -4528,79 +4530,80 @@ private boolean equalityIntersectionExistsForComplexTypes(Set<BType> lhsTypes, S
+                    if (!isAssignable(((BMapType) refLhsMemberType).constraint, symTable.errorType) &&
                             rhsTypes.stream().anyMatch(rhsMemberType -> rhsMemberType.tag == TypeTags.JSON)) {
                         // at this point it is guaranteed that the map is anydata
                         return true;
                     }
 
                     if (rhsTypes.stream().anyMatch(
                             rhsMemberType -> rhsMemberType.tag == TypeTags.RECORD &&
-                                    mapRecordEqualityIntersectionExists((BMapType) lhsMemberType,
-                                                                        (BRecordType) rhsMemberType))) {
+                                    mapRecordEqualityIntersectionExists((BMapType) refLhsMemberType,
+                                            (BRecordType) rhsMemberType))) {
                         return true;
                     }
                     break;
                 case TypeTags.OBJECT:
                 case TypeTags.RECORD:
                     if (rhsTypes.stream().anyMatch(
-                            rhsMemberType -> checkStructEquivalency(rhsMemberType, lhsMemberType) ||
-                                    checkStructEquivalency(lhsMemberType, rhsMemberType))) {
+                            rhsMemberType -> checkStructEquivalency(rhsMemberType, refLhsMemberType) ||
+                                    checkStructEquivalency(refLhsMemberType, rhsMemberType))) {
                         return true;
                     }
 
                     if (rhsTypes.stream().anyMatch(
                             rhsMemberType -> rhsMemberType.tag == TypeTags.RECORD &&
-                                    recordEqualityIntersectionExists((BRecordType) lhsMemberType,
-                                                                     (BRecordType) rhsMemberType))) {
+                                    recordEqualityIntersectionExists((BRecordType) refLhsMemberType,
+                                            (BRecordType) rhsMemberType))) {
                         return true;
                     }
 
                     if (rhsTypes.stream().anyMatch(rhsMemberType -> rhsMemberType.tag == TypeTags.JSON) &&
-                            jsonEqualityIntersectionExists(expandAndGetMemberTypesRecursive(lhsMemberType))) {
+                            jsonEqualityIntersectionExists(expandAndGetMemberTypesRecursive(refLhsMemberType))) {
                         return true;
                     }
 
                     if (rhsTypes.stream().anyMatch(
                             rhsMemberType -> rhsMemberType.tag == TypeTags.MAP &&
                                     mapRecordEqualityIntersectionExists((BMapType) rhsMemberType,
-                                                                        (BRecordType) lhsMemberType))) {
+                                            (BRecordType) refLhsMemberType))) {
```

Comment:
Instead, may be a reference type case here?

Reply:
Updated 904dc9232bd074b2a44a5e941b55d4f5302bb85b